<a href="https://colab.research.google.com/github/Solsmi/challenge_telecomx_latam/blob/main/TelecomX_LATAM.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Diccionario**
- customerID: número de identificación único de cada cliente
- Churn: si el cliente dejó o no la empresa
- gender: género (masculino y femenino)
- SeniorCitizen: información sobre si un cliente tiene o no una edad igual o mayor a 65 años
- Partner: si el cliente tiene o no una pareja
- Dependents: si el cliente tiene o no dependientes
- tenure: meses de contrato del cliente
- PhoneService: suscripción al servicio telefónico
- MultipleLines: suscripción a más de una línea telefónica
- InternetService: suscripción a un proveedor de internet
- OnlineSecurity: suscripción adicional de seguridad en línea
- OnlineBackup: suscripción adicional de respaldo en línea
- DeviceProtection: suscripción adicional de protección del dispositivo
- TechSupport: suscripción adicional de soporte técnico, menor tiempo de espera
- StreamingTV: suscripción de televisión por cable
- StreamingMovies: suscripción de streaming de películas
- Contract: tipo de contrato
- PaperlessBilling: si el cliente prefiere recibir la factura en línea
- PaymentMethod: forma de pago
- Charges.Monthly: total de todos los servicios del cliente por mes
- Charges.Total: total gastado por el cliente

#📌 Extracción

In [93]:
import requests
import pandas as pd

In [94]:
url = 'https://raw.githubusercontent.com/alura-cursos/challenge2-data-science-LATAM/refs/heads/main/TelecomX_Data.json'
response = requests.get(url)
data = response.json()
df = pd.DataFrame(data)
df.head()

customerID Churn                                           customer  \
0  0002-ORFBO    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
1  0003-MKNFE    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
2  0004-TLHLJ   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
3  0011-IGKFF   Yes  {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   
4  0013-EXCHZ   Yes  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   

                                             phone  \
0   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
1  {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
2   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
3   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
4   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   

                                            internet  \
0  {'InternetService': 'DSL', 'OnlineSecurity': '...   
1  {'InternetService': 'DSL', 'OnlineSecurity': '...   
2  {'InternetService': 'Fiber optic', 'OnlineSecu...   
3  {'InternetService': 'Fiber optic', 'OnlineSecu...   
4  {'InternetService': 'Fiber optic', 'OnlineSecu...   

                                             account  
0  {'Contract': 'One year', 'PaperlessBilling': '...  
1  {'Contract': 'Month-to-month', 'PaperlessBilli...  
2  {'Contract': 'Month-to-month', 'PaperlessBilli...  
3  {'Contract': 'Month-to-month', 'PaperlessBilli...  
4  {'Contract': 'Month-to-month', 'PaperlessBilli...

In [95]:
df = pd.json_normalize(data)
df

customerID Churn customer.gender  customer.SeniorCitizen  \
0     0002-ORFBO    No          Female                       0   
1     0003-MKNFE    No            Male                       0   
2     0004-TLHLJ   Yes            Male                       0   
3     0011-IGKFF   Yes            Male                       1   
4     0013-EXCHZ   Yes          Female                       1   
...          ...   ...             ...                     ...   
7262  9987-LUTYD    No          Female                       0   
7263  9992-RRAMN   Yes            Male                       0   
7264  9992-UJOEL    No            Male                       0   
7265  9993-LHIEB    No            Male                       0   
7266  9995-HOTOH    No            Male                       0   

     customer.Partner customer.Dependents  customer.tenure phone.PhoneService  \
0                 Yes                 Yes                9                Yes   
1                  No                  No                9                Yes   
2                  No                  No                4                Yes   
3                 Yes                  No               13                Yes   
4                 Yes                  No                3                Yes   
...               ...                 ...              ...                ...   
7262               No                  No               13                Yes   
7263              Yes                  No               22                Yes   
7264               No                  No                2                Yes   
7265              Yes                 Yes               67                Yes   
7266              Yes                 Yes               63                 No   

     phone.MultipleLines internet.InternetService internet.OnlineSecurity  \
0                     No                      DSL                      No   
1                    Yes                      DSL                      No   
2                     No              Fiber optic                      No   
3                     No              Fiber optic                      No   
4                     No              Fiber optic                      No   
...                  ...                      ...                     ...   
7262                  No                      DSL                     Yes   
7263                 Yes              Fiber optic                      No   
7264                  No                      DSL                      No   
7265                  No                      DSL                     Yes   
7266    No phone service                      DSL                     Yes   

     internet.OnlineBackup internet.DeviceProtection internet.TechSupport  \
0                      Yes                        No                  Yes   
1                       No                        No                   No   
2                       No                       Yes                   No   
3                      Yes                       Yes                   No   
4                       No                        No                  Yes   
...                    ...                       ...                  ...   
7262                    No                        No                  Yes   
7263                    No                        No                   No   
7264                   Yes                        No                   No   
7265                    No                       Yes                  Yes   
7266                   Yes                       Yes                   No   

     internet.StreamingTV internet.StreamingMovies account.Contract  \
0                     Yes                       No         One year   
1                      No                      Yes   Month-to-month   
2                      No                       No   Month-to-month   
3                     Yes                      Yes   Month-to-month   
4                     Yes          

##Inspeccionar Tipos de Datos

In [96]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   customerID                 7267 non-null   object 
 1   Churn                      7267 non-null   object 
 2   customer.gender            7267 non-null   object 
 3   customer.SeniorCitizen     7267 non-null   int64  
 4   customer.Partner           7267 non-null   object 
 5   customer.Dependents        7267 non-null   object 
 6   customer.tenure            7267 non-null   int64  
 7   phone.PhoneService         7267 non-null   object 
 8   phone.MultipleLines        7267 non-null   object 
 9   internet.InternetService   7267 non-null   object 
 10  internet.OnlineSecurity    7267 non-null   object 
 11  internet.OnlineBackup      7267 non-null   object 
 12  internet.DeviceProtection  7267 non-null   object 
 13  internet.TechSupport       7267 non-null   objec

##Inspeccionar Valores Nulos

In [97]:
df.isnull().sum()

customerID                   0
Churn                        0
customer.gender              0
customer.SeniorCitizen       0
customer.Partner             0
customer.Dependents          0
customer.tenure              0
phone.PhoneService           0
phone.MultipleLines          0
internet.InternetService     0
internet.OnlineSecurity      0
internet.OnlineBackup        0
internet.DeviceProtection    0
internet.TechSupport         0
internet.StreamingTV         0
internet.StreamingMovies     0
account.Contract             0
account.PaperlessBilling     0
account.PaymentMethod        0
account.Charges.Monthly      0
account.Charges.Total        0
dtype: int64

## Analizar Valores Únicos en Columnas Tipo *Object*

---



Para las columnas de tipo 'object', examinar los valores únicos para comprender las categorías existentes e identificar posibles inconsistencias o valores inesperados.


In [98]:
object_columns = df.select_dtypes(include='object').columns

for col in object_columns:
    print(f"Column: {col}")
    value_counts = df[col].value_counts()
    print(value_counts)

    # destacar valores vacios o espacios en blanco
    empty_values_count = df[col].apply(lambda x: isinstance(x, str) and x.strip() == '').sum()
    if empty_values_count > 0:
        print(f"\033[1m    (Observacion: Se encontraron {empty_values_count} valores vacíos o de solo espacios en blanco)\033[0m") # ANSI escape codes for bold text

    print("\n" + "-" * 30 + "\n")

Column: customerID
customerID
9995-HOTOH    1
0002-ORFBO    1
0003-MKNFE    1
9970-QBCDA    1
9968-FFVVH    1
             ..
0014-BMAQU    1
0013-SMEOE    1
0013-MHZWF    1
0013-EXCHZ    1
0011-IGKFF    1
Name: count, Length: 7267, dtype: int64

------------------------------

Column: Churn
Churn
No     5174
Yes    1869
        224
Name: count, dtype: int64
    (Observacion: Se encontraron 224 valores vacíos o de solo espacios en blanco)

------------------------------

Column: customer.gender
customer.gender
Male      3675
Female    3592
Name: count, dtype: int64

------------------------------

Column: customer.Partner
customer.Partner
No     3749
Yes    3518
Name: count, dtype: int64

------------------------------

Column: customer.Dependents
customer.Dependents
No     5086
Yes    2181
Name: count, dtype: int64

------------------------------

Column: phone.PhoneService
phone.PhoneService
Yes    6560
No      707
Name: count, dtype: int64

------------------------------

Column: ph

El análisis de los valores únicos para las columnas de tipo 'object' ha revelado algunas inconsistencias:

- **Churn**: Además de 'No' y 'Yes', se encontraron 224 entradas con un valor vacío `''`. Esto podría indicar datos faltantes o un problema en la recolección de datos que necesita ser abordado.
- **account.Charges.Total**: Se identificaron 11 entradas con un valor vacío `''`. Dado que esta columna debería ser numérica (total de cargos), estos valores vacíos impedirán el análisis numérico y deben ser tratados. También, la columna `account.Charges.Total` se identificó como tipo `object` en el `df.info()`, lo que sugiere que los valores vacíos podrían estar causando que no se convierta a un tipo numérico adecuado. Esto requerirá una limpieza y conversión de tipo de datos.

##Valores vacíos o en blanco

In [99]:
df.apply(lambda x: x.astype(str).str.strip() == '').sum()

customerID                     0
Churn                        224
customer.gender                0
customer.SeniorCitizen         0
customer.Partner               0
customer.Dependents            0
customer.tenure                0
phone.PhoneService             0
phone.MultipleLines            0
internet.InternetService       0
internet.OnlineSecurity        0
internet.OnlineBackup          0
internet.DeviceProtection      0
internet.TechSupport           0
internet.StreamingTV           0
internet.StreamingMovies       0
account.Contract               0
account.PaperlessBilling       0
account.PaymentMethod          0
account.Charges.Monthly        0
account.Charges.Total         11
dtype: int64

#🔧 Transformación

##Eliminar valores vacíos de Churn

In [100]:
df = df[df['Churn'] != '']
df.reset_index(drop=True, inplace=True)


In [101]:
churn_counts = df['Churn'].value_counts()
print(churn_counts)
print(f"\nSuma Total: {churn_counts.sum()}")

Churn
No     5174
Yes    1869
Name: count, dtype: int64

Suma Total: 7043


##Eliminar cadenas vacías y espacios en blanco de Charges.Total

In [102]:
df = df[df['account.Charges.Total'] != '']
df = df[df['account.Charges.Total'].str.strip() != '']
df.reset_index(drop=True, inplace=True)
df['account.Charges.Total'] = pd.to_numeric(df['account.Charges.Total'])

In [103]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7032 entries, 0 to 7031
Data columns (total 21 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   customerID                 7032 non-null   object 
 1   Churn                      7032 non-null   object 
 2   customer.gender            7032 non-null   object 
 3   customer.SeniorCitizen     7032 non-null   int64  
 4   customer.Partner           7032 non-null   object 
 5   customer.Dependents        7032 non-null   object 
 6   customer.tenure            7032 non-null   int64  
 7   phone.PhoneService         7032 non-null   object 
 8   phone.MultipleLines        7032 non-null   object 
 9   internet.InternetService   7032 non-null   object 
 10  internet.OnlineSecurity    7032 non-null   object 
 11  internet.OnlineBackup      7032 non-null   object 
 12  internet.DeviceProtection  7032 non-null   object 
 13  internet.TechSupport       7032 non-null   objec

##Estandarizacion español

In [104]:
columnas = {    # Diccionario con los nuevos nombres de las columnas.
    'customerID': 'id_cliente',
    'Churn': 'abandono',
    'customer.gender': 'genero',
    'customer.SeniorCitizen': 'tiene +60',
    'customer.Partner': 'posee_pareja',
    'customer.Dependents': 'posee_dependientes',
    'customer.tenure': 'tiempo_contrato',
    'phone.PhoneService': 'servicio_telefono',
    'phone.MultipleLines': 'multiples_lineas',
    'internet.InternetService': 'tipo_internet',
    'internet.OnlineSecurity': 'seguridad_online',
    'internet.OnlineBackup': 'backup_online',
    'internet.DeviceProtection': 'proteccion_dispositivo',
    'internet.TechSupport': 'soporte_tecnico',
    'internet.StreamingTV': 'streaming_tv',
    'internet.StreamingMovies': 'streaming_peliculas',
    'account.Contract': 'tipo_contrato',
    'account.PaperlessBilling': 'factura_digital',
    'account.PaymentMethod': 'metodo_pago',
    'account.Charges.Monthly': 'cargo_mensual',
    'account.Charges.Total': 'cargo_total',

}
df = df.rename(columns= columnas)
df

id_cliente abandono  genero  tiene +60 posee_pareja posee_dependientes  \
0     0002-ORFBO       No  Female          0          Yes                Yes   
1     0003-MKNFE       No    Male          0           No                 No   
2     0004-TLHLJ      Yes    Male          0           No                 No   
3     0011-IGKFF      Yes    Male          1          Yes                 No   
4     0013-EXCHZ      Yes  Female          1          Yes                 No   
...          ...      ...     ...        ...          ...                ...   
7027  9987-LUTYD       No  Female          0           No                 No   
7028  9992-RRAMN      Yes    Male          0          Yes                 No   
7029  9992-UJOEL       No    Male          0           No                 No   
7030  9993-LHIEB       No    Male          0          Yes                Yes   
7031  9995-HOTOH       No    Male          0          Yes                Yes   

      tiempo_contrato servicio_telefono  multiples_lineas tipo_internet  \
0                   9               Yes                No           DSL   
1                   9               Yes               Yes           DSL   
2                   4               Yes                No   Fiber optic   
3                  13               Yes                No   Fiber optic   
4                   3               Yes                No   Fiber optic   
...               ...               ...               ...           ...   
7027               13               Yes                No           DSL   
7028               22               Yes               Yes   Fiber optic   
7029                2               Yes                No           DSL   
7030               67               Yes                No           DSL   
7031               63                No  No phone service           DSL   

     seguridad_online backup_online proteccion_dispositivo soporte_tecnico  \
0                  No           Yes                     No             Yes   
1                  No            No                     No              No   
2                  No            No                    Yes              No   
3                  No           Yes                    Yes              No   
4                  No            No                     No             Yes   
...               ...           ...                    ...             ...   
7027              Yes            No                     No             Yes   
7028               No            No                     No              No   
7029               No           Yes                     No              No   
7030              Yes            No                    Yes             Yes   
7031              Yes           Yes                    Yes              No   

     streaming_tv streaming_peliculas   tipo_contrato factura_digital  \
0             Yes                  No        One year             Yes   
1              No                 Yes  Month-to-month              No   
2              No                  No  Month-to-month             Yes   
3             Yes                 Yes  Month-to-month             Yes   
4             Yes                  No  Month-to-month             Yes   
...           ...                 ...             ...             ...   
7027           No                  No        One year              No   
7028           No                 Yes  Month-to-month             Yes   
7029           No                  No  Month-to-month             Yes   
7030           No                 Yes        Two year              No   
7031          Yes                 Yes        Two year              No   

           metodo_pago  cargo_mensual  cargo_total  
0         Mailed check          65.60       593.30  
1         Mailed check          59.90       542.40  
2     Electronic check          73.90       280.85  
3     Electronic check          98.00      1237.85  
4         Mailed check          83.90       267.40  
...         

In [105]:
df['abandono'] = df['abandono'].replace({'Yes': 'Sí'})
df['genero'] = df['genero'].replace({'Female': 'Femenino', 'Male': 'Masculino'})
df['posee_pareja'] = df['posee_pareja'].replace({'Yes': 'Sí'})
df['posee_dependientes'] = df['posee_dependientes'].replace({'Yes': 'Sí'})
df['servicio_telefono'] = df['servicio_telefono'].replace({'Yes': 'Sí'})
df['multiples_lineas'] = df['multiples_lineas'].replace({'Yes': 'Sí', 'No phone service': 'Sin servicio de teléfono'})
df['tipo_internet'] = df['tipo_internet'].replace({'Fiber optic': 'Fibra óptica'})
df['seguridad_online'] = df['seguridad_online'].replace({'Yes': 'Sí', 'No internet service': 'Sin servicio de internet'})
df['backup_online'] = df['backup_online'].replace({'Yes': 'Sí', 'No internet service': 'Sin servicio de internet'})
df['proteccion_dispositivo'] = df['proteccion_dispositivo'].replace({'Yes': 'Sí', 'No internet service': 'Sin servicio de internet'})
df['soporte_tecnico'] = df['soporte_tecnico'].replace({'Yes': 'Sí', 'No internet service': 'Sin servicio de internet'})
df['streaming_tv'] = df['streaming_tv'].replace({'Yes': 'Sí', 'No internet service': 'Sin servicio de internet'})
df['streaming_peliculas'] = df['streaming_peliculas'].replace({'Yes': 'Sí', 'No internet service': 'Sin servicio de internet'})
df['tipo_contrato'] = df['tipo_contrato'].replace({'One year': 'Anual', 'Month-to-month': 'Mensual', 'Two year': 'Bienal'})
df['factura_digital'] = df['factura_digital'].replace({'Yes': 'Sí'})
df['metodo_pago'] = df['metodo_pago'].replace({'Mailed check': 'Cheque por correo', 'Electronic check': 'Cheque electrónico', 'Bank transfer (automatic)': 'Transferencia bancaria (automática)', 'Credit card (automatic)': 'Tarjeta de crédito (automática)'})

In [106]:
pd.set_option('display.max_columns', None)  #ver todos los datos de las columnas

In [107]:
df.head()

id_cliente abandono     genero  tiene +60 posee_pareja posee_dependientes  \
0  0002-ORFBO       No   Femenino          0           Sí                 Sí   
1  0003-MKNFE       No  Masculino          0           No                 No   
2  0004-TLHLJ       Sí  Masculino          0           No                 No   
3  0011-IGKFF       Sí  Masculino          1           Sí                 No   
4  0013-EXCHZ       Sí   Femenino          1           Sí                 No   

   tiempo_contrato servicio_telefono multiples_lineas tipo_internet  \
0                9                Sí               No           DSL   
1                9                Sí               Sí           DSL   
2                4                Sí               No  Fibra óptica   
3               13                Sí               No  Fibra óptica   
4                3                Sí               No  Fibra óptica   

  seguridad_online backup_online proteccion_dispositivo soporte_tecnico  \
0               No            Sí                     No              Sí   
1               No            No                     No              No   
2               No            No                     Sí              No   
3               No            Sí                     Sí              No   
4               No            No                     No              Sí   

  streaming_tv streaming_peliculas tipo_contrato factura_digital  \
0           Sí                  No         Anual              Sí   
1           No                  Sí       Mensual              No   
2           No                  No       Mensual              Sí   
3           Sí                  Sí       Mensual              Sí   
4           Sí                  No       Mensual              Sí   

          metodo_pago  cargo_mensual  cargo_total  
0   Cheque por correo           65.6       593.30  
1   Cheque por correo           59.9       542.40  
2  Cheque electrónico           73.9       280.85  
3  Cheque electrónico           98.0      1237.85  
4   Cheque por correo           83.9       267.40

In [108]:
df['cuentas_diarias'] = df['cargo_mensual'] / 30
df.head()

id_cliente abandono     genero  tiene +60 posee_pareja posee_dependientes  \
0  0002-ORFBO       No   Femenino          0           Sí                 Sí   
1  0003-MKNFE       No  Masculino          0           No                 No   
2  0004-TLHLJ       Sí  Masculino          0           No                 No   
3  0011-IGKFF       Sí  Masculino          1           Sí                 No   
4  0013-EXCHZ       Sí   Femenino          1           Sí                 No   

   tiempo_contrato servicio_telefono multiples_lineas tipo_internet  \
0                9                Sí               No           DSL   
1                9                Sí               Sí           DSL   
2                4                Sí               No  Fibra óptica   
3               13                Sí               No  Fibra óptica   
4                3                Sí               No  Fibra óptica   

  seguridad_online backup_online proteccion_dispositivo soporte_tecnico  \
0               No            Sí                     No              Sí   
1               No            No                     No              No   
2               No            No                     Sí              No   
3               No            Sí                     Sí              No   
4               No            No                     No              Sí   

  streaming_tv streaming_peliculas tipo_contrato factura_digital  \
0           Sí                  No         Anual              Sí   
1           No                  Sí       Mensual              No   
2           No                  No       Mensual              Sí   
3           Sí                  Sí       Mensual              Sí   
4           Sí                  No       Mensual              Sí   

          metodo_pago  cargo_mensual  cargo_total  cuentas_diarias  
0   Cheque por correo           65.6       593.30         2.186667  
1   Cheque por correo           59.9       542.40         1.996667  
2  Cheque electrónico           73.9       280.85         2.463333  
3  Cheque electrónico           98.0      1237.85         3.266667  
4   Cheque por correo           83.9       267.40         2.796667

#📊 Carga y análisis

#📄Informe final